scripts/manila-share-sync.py

```diff
@@ -193,6 +199,66 @@ def sync_share_size(self, shares, dry_run=True):
                         self.set_share_size(share_id, vsize)
                         self.MANILA_SYNC_SHARE_SIZE_COUNTER.inc()
 
+    def sync_share_state(self, shares, dry_run=True):
+        """ Deal with share in stuck states for more than 15 minutes """
+        msg = "ManilaSyncShareState: share=%s, instance=%s status=%s"
+        msg_dry_run = "Dry run: " + msg
+        for share in shares:
+            share_status = share['status']
+            if share_status not in ['creating', 'error_deleting', 'deleting']:
+                continue
+
+            share_id = share['id']
+            instances = self._query_share_instances(share_id)
+            if len(instances) == 0:
+                continue
+            elif len(instances) == 1:
+                instance = instances[0]
+                instance_updated_at = instance['updated_at']
+                if instance_updated_at is not None:
+                    if is_utcts_recent(instance_updated_at, 900):
+                        continue
+                else:
+                    continue
+
+                instance_id = instance['id']
+                instance_status = instance['status']
+                if dry_run:
+                    log.info(msg_dry_run, share_id, instance_id, instance_status)
+                    continue
+                log.info(msg, share_id, instance_id, instance_status)
+                if share_status == 'error_deleting':
+                    self.share_force_delete(share_id)
+                    continue
+                self.share_reset_state(share_id, 'error')
+                if share_status == 'deleting':
+                    self.share_delete(share_id)
+            else:
+                for instance in instances:
+                    instance_updated_at = instance['updated_at']
+                    if instance_updated_at is not None:
+                        if is_utcts_recent(instance_updated_at, 900):
+                            continue
+                    else:
+                        continue
+
+                    if instance.get('replica_state', None) in ['active', None]:
+                        continue
+
+                    instance_id = instance['id']
+                    instance_status = instance['status']
+                    if dry_run:
+                        log.info(msg_dry_run, share_id, instance_id, instance_status)
+                        continue
+                    log.info(msg, share_id, instance_id, instance_status)
+                    if instance_status == 'error_deleting':
+                        self.share_instance_force_delete(instance_id)
+                        continue
+                    self.share_instance_reset_state(instance_id, 'error')
```

Comment:
looks like instance status is set to 'error' only when it is stuck in "writing", isn't it? I would suggest to group this line and the if statements on line 254 and line 258 into a if-else block, to improve readability.

Reply:
When error deleting -> force delete
When deleting -> move to error state and then delete
when creating -> move to error state.

so set to error when both 'creating' and 'deleting'

Reply:
ok, then I recommend to use if-else block
if instance_status == 'error_deleting'
elif instance_status == 'deleting'
else

it does not matter to write 'move to error state' multiple times, readability is more important.

Reply:
At the top of loop we consider only shares in creating/deleting/error_deleting state.

so if share instance is in error_deleting its force deleted
so if share instance is in deleting state, it set to error and deleted.
so if share instance is in creating state, it set to error.

once its set to error, the share wont be considered in next iteration. Is this what you are referring multiple times ?

Logic is same as below
https://github.wdf.sap.corp/d063130/zorro/blob/2ccf81f4cd871f4b3262efe1d4cdf08ab6383cf9/screwdriver_ext_manila/cmds/manila.py#L135

Reply:
![image](https://user-images.githubusercontent.com/90381079/163018840-de871cc4-6c4f-4da0-bdd9-e0721861c48b.png)

211-236 is necessary. Take simple example.
create share - it will create one active replica. Try to delete replica it will fail. We _must_ delete share and not replica.
please check attached image, how share replica delete fails in this case.

secondly the code for retrival share instances is not almost same, its retriving share instances for given share. In share get list, I am considering only active replica state based share (i.e. each share will have single entry in returned list)
And then I process all shares instances of that share. This way code is clear to understand and maintain I believe.

Reply:
ok, I see your points splitting the instances. I will take a look at the details

Reply:
Positive point I find in this PR is the usages of manilaclient, but generally I think the quality of `sync_share_state` and `sync_share_snapshot_state` should be improved before merge. 

* The three states `error_deleting`, `deleting` and `creating` are processed in a mixing way, which leads to using `continue` often. 
*  I understand the reason why you have implemented them, but I still think lines 211-236 and function `query_share_instances` are redundant. One iteration of the shares list should be enough, because all the needed info are already in the `shares` list. And for the problem of deleting share/replica, I think for example, can be solved by using try-except block. We know the exception when deleting the single replica of a share.